ddl/ddl_api.go

```diff
@@ -3383,200 +3383,143 @@ func checkModifyCharsetAndCollation(toCharset, toCollate, origCharset, origColla
-			}
-			return unsupportedMsg, errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-		}
-		// Floating-point and fixed-point types also can be UNSIGNED. As with integer types, this attribute prevents
-		// negative values from being stored in the column. Unlike the integer types, the upper range of column values
-		// remains the same.
-		if to.Flen != origin.Flen || to.Decimal != origin.Decimal || mysql.HasUnsignedFlag(to.Flag) != mysql.HasUnsignedFlag(origin.Flag) {
-			msg := fmt.Sprintf("decimal change from decimal(%d, %d) to decimal(%d, %d)", origin.Flen, origin.Decimal, to.Flen, to.Decimal)
-			return msg, errUnsupportedModifyColumn.GenWithStackByArgs(msg)
-		}
-	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeDuration, mysql.TypeYear:
-		switch origin.Tp {
-		case mysql.TypeDuration:
-			switch to.Tp {
-			case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
-				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-			}
-		case mysql.TypeDate:
-			switch to.Tp {
-			case mysql.TypeDatetime, mysql.TypeTimestamp:
-				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-			}
-		case mysql.TypeTimestamp:
-			switch to.Tp {
-			case mysql.TypeDuration, mysql.TypeDatetime:
-				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-			}
-		case mysql.TypeDatetime:
-			switch to.Tp {
-			case mysql.TypeTimestamp:
-				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-			}
-		case mysql.TypeYear:
-			switch to.Tp {
-			case mysql.TypeDuration:
-				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-			case mysql.TypeYear:
-			default:
-				return "", ErrTruncatedWrongValue.GenWithStack("banned conversion that must fail")
+
+		if origin.Tp == mysql.TypeNewDecimal {
+			// Floating-point and fixed-point types also can be UNSIGNED. As with integer types, this attribute prevents
+			// negative values from being stored in the column. Unlike the integer types, the upper range of column values
+			// remains the same.
+			if to.Flen != origin.Flen || to.Decimal != origin.Decimal || mysql.HasUnsignedFlag(to.Flag) != mysql.HasUnsignedFlag(origin.Flag) {
+				msg := fmt.Sprintf("decimal change from decimal(%d, %d) to decimal(%d, %d)", origin.Flen, origin.Decimal, to.Flen, to.Decimal)
+				return true, msg, errUnsupportedModifyColumn.GenWithStackByArgs(msg)
 			}
 		}
-		switch to.Tp {
-		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeDuration:
-			skipSignCheck = true
-			skipLenCheck = true
-		case mysql.TypeYear:
-			if origin.Tp != mysql.TypeYear {
-				return "", ErrWarnDataOutOfRange.GenWithStack("banned conversion that must fail")
-			}
-			skipSignCheck = true
-			skipLenCheck = true
-		default:
-			return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
+
+		needReorg, reason := needReorgToChange(origin, to)
+		if !needReorg {
+			return
```

Comment:
although go support this syntax, if you return explicit, it will make code more clear